src/critics/local_goal_critic.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022 @artofnothingness Alexey Budyakov, Samsung Research
+#include "mppic/critics/local_goal_critic.hpp"
+
+#include <xtensor/xmath.hpp>
+#include <xtensor/xsort.hpp>
+
+namespace mppi::critics
+{
+
+void LocalGoalCritic::initialize()
+{
+  auto getParam = parameters_handler_->getParamGetter(name_);
+
+  getParam(consider_angle_, "consider_angle_", true);
+  getParam(angle_offset_, "angle_offset", 6);
+  getParam(angle_cost_power_, "angle_cost_power", 1);
+  getParam(angle_cost_weight_, "angle_cost_weight", 3.0);
+
+  getParam(consider_distance_, "consider_distance", true);
+  getParam(distance_goal_count_, "disance_goal_count", 2);
+  getParam(distance_offset_, "distance_offset", 6);
+  getParam(distance_cost_power_, "distance_cost_power", 2);
+  getParam(distance_cost_weight_, "distance_cost_weight", 3.0);
+
+  getParam(stop_usage_path_reached_ratio_, "stop_usage_path_reached_ratio", 0.35);
+}
+
+void LocalGoalCritic::score(models::CriticFunctionData & data)
+{
+  if(!enabled_) {
+    return;
+  }
+
+  auto path_points = xt::view(
+    data.path, xt::all(), xt::range(0, 2));
+
+  auto find_furthest_point = [&]() {
```

Comment:
I believe this process can be done more efficiently

Get the last set of points in the trajectories (`last_points_ext`, though I'm not sure what "ext" means here), find the max distance of them so we have a single number for the max length. Then simply march through the path starting at the robot pose and find the `idx` on the path that is larger than the max length you previously found. 

That would save you iterating over every trajectory over every path point, now you're just iterating over the path 1 time and finding using tensor optimizations the single value you care about before comparison

Reply:
Also, is there a reason not to just break this out as a separate function?

Reply:
I would suggest optimize this in PR with reference critic optimization, change lambda to function